File: Dante2/Module2.05/Zadanie5.04/main.c
#include <stdio.h>
#include <stdlib.h>

struct matrix_t
{
    int **ptr;
    int width;
    int height;
};

int matrix_create(struct matrix_t *m, int width, int height);
int matrix_read(struct matrix_t *m);
void matrix_display(const struct matrix_t *m);
void matrix_destroy(struct matrix_t *m);

int main()
{
    int error;
    int width, height;

    printf("Enter: ");
    error = scanf("%i %i",&width, &height);
    if(error != 2)
    {
        printf("Incorrect input");
        return 1;
    }
    if(width <= 0 || height <= 0)
    {
        printf("Incorrect input data");
        return 2;
    }

    struct matrix_t m;
    error = matrix_create(&m, width, height);
    if(error == 2)
    {
        printf("Failed to allocate memory");
        return 8;
    }
    if(error == 1)
    {
        printf("Incorrect input data");
        return 2;
    }

    printf("Enter:\n");
    error = matrix_read(&m);
    if(error == 1)
    {
        printf("Incorrect input data");
        matrix_destroy(&m);
        return 2;
    }
    if(error == 2)
    {
        printf("Incorrect input");
        matrix_destroy(&m);
        return 1;
    }

    matrix_display(&m);
    matrix_destroy(&m);
    return 0;
}

int matrix_read(struct matrix_t *m)
{
    if(m == NULL || m->ptr == NULL || m->height <= 0 || m->width <= 0)
    {
        return 1;
    }

    int error;

    for(int i = 0; i < m->height; i++)
    {
        for(int j = 0; j < m->width; j++)
        {
            error = scanf("%i",(*(m->ptr + i) + j));
            if(!error)
            {
                matrix_destroy(m);
                return 2;
            }
        }
    }

    return 0;
}

void matrix_display(const struct matrix_t *m)
{
    if(m == NULL || m->height <= 0 || m->width <= 0 || m->ptr == NULL)
    {
        return;
    }

    for (int i = 0; i < m->height; i++)
    {
        for (int j = 0; j < m->width; j++)
        {
            printf("%i ", *(*(m->ptr + i) + j));
        }
        printf("\n");
    }
}

void matrix_destroy(struct matrix_t *m)
{
    if(m == NULL || m->height <= 0 || m->width <= 0 || m->ptr == NULL)
    {
        return;
    }

    for(int i = 0; i < m->height; i++)
    {
        free(*(m->ptr + i));
    }
    free(m->ptr);
    m->ptr = NULL;

    m->height = 0;
    m->width = 0;
}

int matrix_create(struct matrix_t *m, int width, int height)
{
    if(m == NULL || width <= 0 || height <= 0)
    {
        return 1;
    }

    m->ptr = malloc(sizeof(int *) * height);
    if(m->ptr == NULL)
    {
        return 2;
    }

    for(int i = 0; i < height; i++)
    {
        *(m->ptr + i) = malloc(sizeof(int) * width);
        if(*(m->ptr + i) == NULL)
        {
            for (int j = 0; j < i; j++)
            {
                free(*(m->ptr + j));
            }
            free(m->ptr);
            return 2;
        }
    }

    m->width = width;
    m->height = height;

    return 0;
}

File: Dante2/Module2.05/Zadanie5.04/main.rs
use regex::Regex;

#[derive(Default)]
pub struct MatrixT {
    ptr: Vec<Vec<i32>>,
    width: i32,
    height: i32,
}

pub fn matrix_create(m: &mut MatrixT, width: i32, height: i32) -> Result<i32, i32> {
    if height <= 0 || width <= 0 {
        return Err(1);
    }

    m.width = width;
    m.height = height;
    m.ptr = vec![vec![0; width as usize]; height as usize];

    Ok(0)
}

pub fn matrix_read(m: &mut  MatrixT) -> Result<i32, i32> {
    if m.height <= 0 || m.width <= 0 {
        return Err(1);
    }

    let mut input: String = String::new();
    for i in 0..m.height {
        for j in 0..m.width {
            input.clear();
            std::io::stdin().read_line(&mut input).expect("Incorrect input");
            m.ptr[i as usize][j as usize] = input.trim().parse().expect("Incorrect parse");
        }
    }

    Ok(0)
}

pub fn matrix_display(m: & MatrixT) {
    if m.width <= 0 || m.height <= 0 {
        return;
    }

    println!();
    for i in 0..m.height {
        for j in 0..m.width {
            print!("{} ", m.ptr[i as usize][j as usize])
        }
        println!();
    }
    println!();
}


fn main() {
    let mut input: String = String::new();

    println!("Enter data: ");
    std::io::stdin().read_line(&mut input).expect("Incorrect input");
    let re = Regex::new(r"(\d+)\s+(\d+)").expect("Incorrect REGEX");
    if let Some(caps) = re.captures(&input) {
        let width: i32 = caps[1].parse().expect("Incorrect width parse");
        let height: i32 = caps[2].parse().expect("Incorrect height parse");

        if width <= 0 || height <= 0 {
            panic!("Incorrect input data");
        }

        let mut m: MatrixT = Default::default();
        match matrix_create(&mut m, width, height) {
            Err(1) => panic!("Incorrect input"),
            _other => {},
        }

        matrix_display(&m);

        println!("Enter matrix: ");
        match matrix_read(&mut m) {
            Err(1) => panic!("Incorrect matrix in matrix_read"),
            _other => {},
        }

        matrix_display(&m);
    }
}
